org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnDiscover.java

```diff
@@ -0,0 +1,311 @@
+      List<Object> params, IProgressMonitor pm) {
+    Preconditions.checkArgument(params.size() >= 1, "FQN is expected.");
+
+    String fqn = (String) params.get(0);
+
+    if (pm.isCanceled()) {
+      throw new OperationCanceledException();
+    }
+
+    Pair<char[][], char[][]> fqnPair = prepareFqnToSearch(fqn);
+    List<IType> types;
+    try {
+      types =
+          findTypeByFqn(
+              fqnPair.getKey(), fqnPair.getValue(), SearchEngine.createWorkspaceScope(), pm);
+    } catch (JavaModelException e) {
+      throw new RuntimeException(e);
+    }
+
+    if (types.isEmpty()) {
+      throw new RuntimeException("Type with fully qualified name: " + fqn + " was not found");
+    }
+
+    IType type = types.get(0); // TODO we need handle few result! It's temporary solution.
+    if (type.isBinary()) {
+      IClassFile classFile = type.getClassFile();
+      int libId = classFile.getAncestor(IPackageFragmentRoot.PACKAGE_FRAGMENT_ROOT).hashCode();
+      return Collections.singletonList(new ResourceLocationParameters(fqn, libId));
+    } else {
+      return Collections.singletonList(
+          new ResourceLocationParameters(JDTUtils.toURI(type.getCompilationUnit())));
+    }
+  }
+
+  private static List<IType> findTypeByFqn(
+      char[][] packages, char[][] names, IJavaSearchScope scope, IProgressMonitor pm)
+      throws JavaModelException {
+    List<IType> result = new ArrayList<>();
+
+    SearchEngine searchEngine = new SearchEngine();
+    searchEngine.searchAllTypeNames(
+        packages,
+        names,
+        scope,
+        new TypeNameMatchRequestor() {
+          @Override
+          public void acceptTypeNameMatch(TypeNameMatch typeNameMatch) {
+            result.add(typeNameMatch.getType());
+          }
+        },
+        IJavaSearchConstants.WAIT_UNTIL_READY_TO_SEARCH,
+        pm);
+    return result;
+  }
+
+  private static Pair<char[][], char[][]> prepareFqnToSearch(String fqn) {
+    String outerClassFqn = extractOuterClassFqn(fqn);
+    int lastDotIndex = outerClassFqn.trim().lastIndexOf('.');
+
+    char[][] packages;
+    char[][] names;
+    if (lastDotIndex == -1) {
+      packages = new char[0][];
+      names = new char[][] {outerClassFqn.toCharArray()};
+    } else {
+      String packageLine = fqn.substring(0, lastDotIndex);
+      packages = new char[][] {packageLine.toCharArray()};
+
+      String nameLine = fqn.substring(lastDotIndex + 1, outerClassFqn.length());
+      names = new char[][] {nameLine.toCharArray()};
+    }
+    return Pair.of(packages, names);
+  }
+
+  private static String extractOuterClassFqn(String fqn) {
+    // handle fqn in case nested classes
+    if (fqn.contains("$")) {
+      return fqn.substring(0, fqn.indexOf("$"));
+    }
+
+    // handle fqn in case lambda expressions
+    return fqn.contains("$$") ? fqn.substring(0, fqn.indexOf("$$")) : fqn;
+  }
+
+  /**
+   * Searches the given source range of the container for a member that is not the same as the given
```

Comment:
I have no idea what this sentence means. Can you clarify?

Reply:
I am not author of the origin class, so I haven't idea. it is better to remove then.